### ISSUE TITLE
Some checking about generics broken after 1.1.87

### ISSUE THREAD

**Issue by giyyapan** [microsoft/pyright#1191]
**Describe the bug**
After upgrading to 1.1.87, Pyright reports some generic issues incorrectly.

Please see the code and screenshots. In this example, the result of p2.chain() should have the same type  with p2, but an error `Type "P" cannot be assigned to type "TP"` is reported.

```python
from typing import TypeVar

class P:
    def chain(self) -> "P":
        ...

TP = TypeVar("TP", bound=P)

p1: P = P() 
p1 = p1.chain() # no error

def fn(p2: TP)-> TP:
    p2 = p2.chain() # error
    return p2
```
![image](https://user-images.githubusercontent.com/3479453/99814463-fa996880-2b83-11eb-87e5-606e5c97aa4c.png)

I've tried changing class P to:
```python
class P:
    def chain(self):
        return self
```
But still no luck.
![image](https://user-images.githubusercontent.com/3479453/99814816-78f60a80-2b84-11eb-8114-82f0bc5eb6f6.png)

**VS Code extension or command-line**
VSCode 1.1.87

**Additional context**
Please let me know if you need more information.
Thanks!

**Comment by K3UL** [comment]
Hello, here is another snipper, please let me know if this needs a new issue but I think the root cause might be the same. This used to work up to 1.1.86 and now breaks on 1.1.87 with message 
```
Argument of type "str" cannot be assigned to parameter "k" of type "StrEnum" in function "__getitem__"
    "str" is incompatible with "StrEnum"
```

The sample : 
```python
from __future__ import annotations
from enum import Enum
from typing import TypeVar

T = TypeVar('T')


class StrEnum(str, Enum):
    def __str__(self):
        return self.value

    @classmethod
    def from_str(cls: type[StrEnum], value: str) -> T:
        """access enum value from string key"""
        if not value:
            raise ValueError(f"No member found with the name '{value}' in type '{T}'")
        return {v: k for k, v in cls.__members__.items()}[value.upper()]

    @classmethod
    def to_list(cls: type[StrEnum]) -> list[str]:
        """retrieve all enumeration as list of strings"""
        return list(cls.__members__.values())


class Something(StrEnum):
    ONE = "first"
    TWO = "second"


thing = Something.from_str("first")
```

**Comment by erictraut** [comment]
Thanks, and apologies for the regressions. I made a very large change in 1.1.87 relating to the handling of TypeVars. I'll make it a top priority to fix these regressions.

**Comment by erictraut** [comment]
@giyyapan, I think Pyright is doing the right thing for your code sample. It previously did not emit an error in this case, but that was a bug. Its behavior now matches that of mypy in this sample.

@K3UL, likewise, I think Pyright is doing the right thing here. This code sample has multiple problems. First, it's using a type variable `T` in a manner that makes no sense, since it's not bound and used only as the return type of a function. Second, the expression `{v: k for k, v in cls.__members__.items()}` has a type of `Dict[StrEnum, str]`, but you are attempting to index it using a key that is of type `str`. For comparison, mypy also emits errors for this code sample.

**Comment by K3UL** [comment]
@erictraut Thank you indeed the behavior never manifested because so far the enum names match the values, but after your comment I dived into it and indded it was not only wrong but overly complicated compared to the right solution.

Thank you again for looking into it so fast !

**Comment by giyyapan** [comment]
@erictraut, Thanks for your reply and explaination, but I still don't understand why `p2 = p2.chain()` should emit bug in this case. 
Could you please give me some reference about that?

In my usecase, the method `chain` of class `P` is designed and implemented to return it's self, if there's any abuse of TypeVar in my code sample, chould you please provide a correct way to make it work?

**Comment by giyyapan** [comment]
Or is there a workaround for that? We have a bunch of codes rely on that, like:
```python
TQuery = TypeVar("TQuery")

def pagination(q: TQuery, page: int, page_size = 30) -> TQuery:
    return q.offset((page - 1) * page_size).limit(page_size) # error here

users = pagination(User.query.all(), 1)
```
And there's a tons of errors now :cry:

**Comment by erictraut** [comment]
Here's your first example with a few more details filled in. I've defined a small class hierarchy with a base class of `Animal` to help explain the problem.

```python
class Animal:
    pass

class Cow(Animal):
    pass

_TA = TypeVar("_TA", bound=Animal)

def fn(p2: _TA)-> _TA:
    p2 = Animal() # error because p2 is declared as _TA
    return p2
```

The function `fn` declares that it takes an input parameter of type `_TA` and returns a value of the same type. We don't know exactly what type `p2` is, but we know that it is an instance of `Animal` or a subclass thereof. Whatever type it is, the function must return the same type. If we attempt to return an instance of `Animal`, it's a type violation because we can no longer guarantee that the input type matches the return type. If the input type is a `Cow`, the return type (and the symbol `p2`) must remain a `Cow`.

You should be able to work around the issue with your `chain` method by declaring it as follows:

```python
TP = TypeVar("TP", bound="P")

class P:
    def chain(self: TP) -> TP:
        return self
```

This should preserve the type of `self` even if it's a subclass of `P`. I use the word "should" because this case isn't currently working in Pyright. I'll file a separate bug on that issue.

As for your sample involving `TQuery`, that piece of code doesn't make sense to me. You're using an unbound TypeVar in ways that is not allowed. Pyright should be emitting multiple errors for this code. This is part of #1112, which I plan to implement in the coming week.

**Comment by giyyapan** [comment]
@erictraut Thank you very much for your detailed explaination. :smiley:

I changed by code sample according to your advice:
```python
from typing import TypeVar

T = TypeVar("T")

class Animal:
    def chain(self: T) -> T:
        ...

class Cow(Animal):
    pass

T2 = TypeVar("T2", bound=Animal)

def fn(p2: T2)-> T2:
    p2 = p2.chain() # error
    return p2
```
But there's still an error. I also tried changing `T` to `TypeVar("T", bound="Animal")`, didn't work either.
![image](https://user-images.githubusercontent.com/3479453/99883079-d8751880-2c5f-11eb-931b-f3b659576e68.png)

If the key here is about the return type of method `chain` (it can't be a explicit type), then replacing it with a TypeVar should have solved the problem. Am I still not getting it?

As for the `TQuery` doesn't make sense part, that part of code was hand written and simplified for presentation, in the real use case `TQuery` are bound to `Query` class, which is the return type of `User.query.all()`.

Thanks you again!

**Comment by erictraut** [comment]
Yes, this code should work, but it currently doesn't. I mentioned above that I filed a separate bug on this. Here's the bug if you want to track it: #1192

**Comment by giyyapan** [comment]
Cool, thank you very much!
